sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,206 @@
+"""
+class Verifier(object):
+    def __init__(self, sydent):
+        self.sydent = sydent
+        # Cache of server keys. These are cached until the 'valid_until_ts' time
+        # in the result.
+        self.cache = {
+            # server_name: <result from keys query>,
+        }
+
+    @defer.inlineCallbacks
+    def _getEndpointForServer(self, server_name):
+        if ':' in server_name:
+            defer.returnValue(tuple(server_name.rsplit(':', 1)))
+
+        service_name = "%s.%s.%s" % ('_matrix', '_tcp', server_name)
+
+        default = server_name, 8448
+
+        try:
+            answers, _, _ = yield twisted.names.client.lookupService(service_name)
+        except DNSNameError:
+            logger.info("DNSNameError doing SRV lookup for %s - using default", service_name)
+            defer.returnValue(default)
+
+        for answer in answers:
+            if answer.type != twisted.names.dns.SRV or not answer.payload:
+                continue
+
+            # XXX we just use the first
+            logger.info("Got SRV answer: %r / %d for %s", str(answer.payload.target), answer.payload.port, service_name)
+            defer.returnValue((str(answer.payload.target), answer.payload.port))
+
+        logger.info("No valid answers found in response from %s (%r)", server_name, answers)
+        defer.returnValue(default)
+
+    @defer.inlineCallbacks
+    def _getKeysForServer(self, server_name):
+        """Get the signing key data from a home server.
+        """
+
+        if server_name in self.cache:
+            cached = self.cache[server_name]
+            now = int(time.time() * 1000)
+            if cached['valid_until_ts'] > now:
+                defer.returnValue(self.cache[server_name]['verify_keys'])
+
+        host_port = yield self._getEndpointForServer(server_name)
+        logger.info("Got host/port %s/%s for %s", host_port[0], host_port[1], server_name)
+        client = FederationHttpClient(self.sydent)
+        result = yield client.get_json("https://%s:%s/_matrix/key/v2/server/" % host_port)
+        if 'verify_keys' not in result:
+            raise Exception("No key found in response")
+
+        if 'valid_until_ts' in result:
+            # Don't cache anything without a valid_until_ts or we wouldn't
+            # know when to expire it.
+            logger.info("Got keys for %s: caching until %s", server_name, result['valid_until_ts'])
+            self.cache[server_name] = result
+
+        defer.returnValue(result['verify_keys'])
+
+    @defer.inlineCallbacks
+    def verifyServerSignedJson(self, signed_json, acceptable_server_names=None):
+        """Given a signed json object, try to verify any one
+        of the signatures on it
+        XXX: This contains a very noddy version of the home server
+        SRV lookup and signature verification. It forms HTTPS URLs
+        from the result of the SRV lookup which will mean the Host:
+        parameter in the request will be wrong. It only looks at
```

Comment:
isn't the wrong Host header going to mess things up? People will have servers behind proxies which rely on it being right.

Reply:
Yeah, this is taken from the impl I did for user directory searching. I've also realised we actually have another impl of this in sydent for the onbind callbacks (https://github.com/matrix-org/sydent/blob/master/sydent/threepid/bind.py#L91) which has the same shortcoming. I guess now is the time to fix both of them as this presumably is going to start impacting people.